File: 020ValidParentheses.cpp
//1.Easily use stack. O(N)
class Solution {
public:
    bool isValid(string s) {
        if(s.empty()) return true;
        
        map<char,char> temp = {{')','('},{']','['},{'}','{'}};
        
        stack<char> st;
        
        for(char c : s){
            if(st.empty()) st.push(c);
            else{
                if(st.top()==temp[c]) st.pop();
                else st.push(c);  
            }    
        }
        
        return st.empty();
    }
};

File: Algo/Leetcode/020ValidParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        n = len(s)
        i = 0
        st = []
        while i < n:
            if s[i] == ')' and len(st) > 0 and st[-1] == '(':
                st.pop()
            elif s[i] == '}' and len(st) > 0 and st[-1] == '{':
                st.pop()
            elif s[i] == ']' and len(st) > 0 and st[-1] == '[':
                st.pop()

            else:
                st.append(s[i])

            i += 1

        return len(st) == 0